nginx_proxy/data/nginx.conf

```diff
@@ -17,6 +17,10 @@ http {
 
     server_names_hash_bucket_size 64;
 
+    # enable modsecurity
+    modsecurity on;
+    modsecurity_rules_file /etc/nginx/modsec/main.conf;
```

Comment:
Might I suggest only including this section if the user has actually turned "On" or "DetectionOnly" mode? You can either "sed" it into the nginx.conf file or set it using tempio.

Reply:
Having `modsecurity on;` only causes the modsecurity module to be loaded by nginx,  but the `SecRuleEngine` setting in modsecurity.conf is what actually determines if any rules are being processed by modsecurity or not. 

If we leave `modsecurity on;` with `security_mode: Off`, the net effect is the same as if we remove the modsecurity entries entirely from nginx.conf: modsecurity rules are not processed.

However, with `modsecurity on;`, there is the side effect that the modsecurity module is loaded by nginx, even though  it may or may not be processing rules depending on the value of `security_mode` option. 

My question would be is there is any concern with nginx loading the modsecurity module if modsecurity is not actually processing rules (i.e. `security_mode: Off`)?

P.S. Thanks for taking the time to review this. I appreciate the feedback and discussion.

Reply:
Because we're running in a state where we're restarting the container each time with a newly built configuration, the only real cost is system memory to only turn on the module when it's in On or DetectionOnly mode. The purist in me doesn't like enabling a module in a configuration and not using it... but I'm used to working in embedded systems. :)

Alternately, maybe adopting an approach similar to the "customize" section...

security:
  active: true
  mode: (On|DetectionOnly[default])

When the section is set active, drop in the modsecurity stanza into nginx.conf and it adopts DetectionOnly with no "mode" set. Otherwise, the user can define the mode manually. This approach would also allow you to customize modsecurity later (custom rule folders? logging destinations?).

Reply:
I like the flexibility this would add to allow for future customizations if needed, with out really adding much complexity initially. Plus the added benefit of not loading modsecurity into nginx at all if it's not being used. Probably, would only have a small reduction in the memory footprint, but should eliminate any possibility of unintended side effects from modsecurity when it is not used.

Reply:
Updated to use this format and tested good on my local. Thanks for the suggestion.

Reply:
Looks great! I'd imagine that it also means it's safer for @frenck to ship as it'll basically run like the "normal" addon by default and only load and engage when the user is ready to configure.

I note that you've got three "exception" rules incorporated (I too had figured Node-RED would be a difficult one). How many addons/integrations have you tested with so far?

Reply:
Good call out. A couple thoughts around this before I answer your question:

1. It would be impossible to find every false positive that might be detected by ModSecurity while using Home Assistant due to the hundreds of "official" integrations, plus essentially unlimited community content. Really the only things that potentially **need** an exception would be official integrations and add-ons that **require** remote access to function (like, Google Assistant, Alexa, etc).
2. Even then, this an exception is only absolutely required if all of the following are true:
   - Built-in, integration, or add-on functionality is detected as a false positive by the default CRS ruleset
   - The integration or add-on requires continuous remote access to function
3. If there is no false positive, then there is no exception required
4. If there is a false positive, and the integration/add-on doesn't require remote access, or only needs one-off remote access (for example, initial registration only, after which it functions locally) i've added some recommendations for local workarounds to DOCS.md
```md
   1. Avoid requests being processed through ModSecurity by accessing Home Assistant using the internal connection URL (e.g. homeassistant.local:8123), then retry the blocked action.
   2. Temporarily disable ModSecurity by changing `security.mode` to `DetectionOnly`, restart the NGINX Home Assistant SSL Proxy add-on, then retry the blocked action. Afterwards, re-enable ModSecurity by changing `security.mode` to `On` and restart the add-on again.
 ```
5. Based on that criteria, of the 3 exceptions I added, the only one that I think is **required** is `SQLi false positive for wildcards in add-on configurations`. This is because you could end up in a situation where you only have remote access to your system, enable ModSecurity, and then cannot disable it because the configuration options get detected SQLi false positive, so the configuration change fails to save, making both workarounds unavailable.
6. The other 2 exceptions I added more due to the "annoyance factor", but maybe they should be removed to not set a precedent?
   - For Node Red: I usually connect using homeassistant.local, but for some reason I was connected to my external address (probably because I was working on this addon). I made some changes to a flow, went to deploy, deploy got blocked by ModSecurity (because your literally POSTing code in the payload, so not unexpected to get flagged). Go to a new browser tab to configure the settings to put ModSecurity in DetectionOnly mode, restart the add-on. Node Red tab loses connection due to nginx addon restart, and when it reconnects, the page reloads and lost my flow changes. Node Red is a "Community Add-on" and does not require remote access to function, so it does not meet my previously stated criteria for required exceptions, so maybe this exclusion should be removed? Or does the higher impact factor in, since you can't simply retry the request without redoing your work?
   - For "Allow integrations to be deleted": This is built-in functionality, so the first criteria of "built-in functionality" is met, but deleting an integration does not require remote access and can be done through local connection, or easily retried after temporarily disabling ModSecurity. I added this one because there is no error message when deleting the integration is blocked; it fails silently. So it took me a few times of hitting the "delete" button and nothing happening before I went to check the NGINX add-on logs to find the block message there.
7. Both of those problems, they can be easily avoided by operating primarily over the local connection, so my recommendation would be if you want to be able have 100% functionality to remotely manage Home Assistant over remote web access at anytime or any place, then you should leave ModSecurity disabled, which would make you no worse off than your currently are. Alternatively, use a VPN connection for most remote management, and use the remote web access for only the integrations/add-ons that require it.
8. However, my impression is that most people only use remote access for minimum necessary functionality of integrations require it. And for those people, they could opt-in to an additional layer of security without any impact to functionality.
9. As you can see from the comment in the exlusions file 'User specified exclusions can be added by using the `security_exclusions_file`', I got a little carried away since there is no `security_exclusions_file` option, and I think 99.9% of people wouldn't write their own ModSecurity rules even if it was an option. But if it turns out this customization is needed, it will be easier to implement due to the adjusts we made to the add-on options. 🎉

With all that being said I can only test remote access for integrations that I use, but I have no false positives with them from ModSecurity:
   - Mobile App
   - SmartThings
   - Withings

I think SmartThings and Withings just use the remote access for external webhook callback, and I think integrations that follow this same behavior have low risk of false positives.

Would be great if somebody could test with the cloud-based assistants, e.g. Google, Alexa. Those are the ones I would think have a higher potential for false positives where a workaround would not be suitable.
I could create an account and configure an integration for those, but I don't know that just setting up the integration is a great test for these without actively using them.

Reply:
Good observations. I'm obviously not saying it's possible to cover every possible integration, but it might be good to cover the HA addons that are part of the Core at the very least - the ones that use the _Ingress_ functions to present their own WebUI. Those are the most-often-used ones and might generate issues. In particular, I'm thinking of the Zwave addons and maybe the Conbee one.

I also agree that very few people are going to want to make their own exclusions, but it's probably something that will occur with advanced users. You might need to cross that bridge (the "security_exclusions" one) sooner than you think. I'd suggest adopting the same approach as the NGINX customizations one in this same addon - but obviously, that can come later.

I use the Nobu Casa functionality for Alexa/Google Assistant functionality which does not require direct external access (uses SNITun) so I'm not able to test the cloud-based assistants. But that's a good point. 

That said, this is an awesome addition to the NGINX addon. I hope we see it integrated.

Reply:
I think there is an argument that could be made for disabling ModSecurity for ingress, since it's protected by Home Assistants authentication. And if a bad actor is able to authenticate to your Home Assistant instance, then ModSecurity isn't going to help you at that point. On the other hand, it's nice to have it as a layer of security, because you never know when a vulnerability could be found.

As far as maintainability, I could see a community currated exclusions list that supports non-core functionality and advanced use cases that people could optionally include in their own instances. then only the minimum necessary (based on the criteria above) ship with the addon.

The key thing to keep on mind is that it would be opt-in. And if you do opt-in and you find it too disruptive for your particular set of integrations, add-ons, and remote management preference, you can opt back out and be no worse off than you are currently.